### ISSUE TITLE
Twisted Web wsgi container often calls Request.finish after connection is closed

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4025]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4025|
|Type|defect|
|Created|2009-09-14 19:30:06Z|
|Branch|https://github.com/twisted/twisted/tree/wsgi-conn-lost-4025|

This results in an unhandled exception:

```
2009-09-14 15:28:38-0400 [-] Unhandled Error
        Traceback (most recent call last):
          File "/srv/trac/.local/lib/python2.5/site-packages/twisted/application/app.py", line 445, in startReactor
            self.config, oldstdout, oldstderr, self.profiler, reactor)
          File "/srv/trac/.local/lib/python2.5/site-packages/twisted/application/app.py", line 348, in runReactorWithLogging
            reactor.run()
          File "/srv/trac/.local/lib/python2.5/site-packages/twisted/internet/base.py", line 1166, in run
            self.mainLoop()
          File "/srv/trac/.local/lib/python2.5/site-packages/twisted/internet/base.py", line 1175, in mainLoop
            self.runUntilCurrent()
        --- <exception caught here> ---
          File "/srv/trac/.local/lib/python2.5/site-packages/twisted/internet/base.py", line 752, in runUntilCurrent
            f(*a, **kw)
          File "/srv/trac/.local/lib/python2.5/site-packages/twisted/web/wsgi.py", line 299, in wsgiFinish
            self.request.finish()
          File "/srv/trac/.local/lib/python2.5/site-packages/twisted/web/http.py", line 828, in finish
            "Request.finish called on a request after its connection was lost; "
        exceptions.RuntimeError: Request.finish called on a request after its connection was lost; use Request.notifyFinish to keep track of this.
```

<details><summary>Searchable metadata</summary>

```
trac-id__4025 4025
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_wsgi_conn_lost_4025 branches/wsgi-conn-lost-4025
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1252956606000000 1252956606000000
changetime__1253524447000000 1253524447000000
version__None None
owner__therve therve

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [27336]) Branching to 'wsgi-conn-lost-4025'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Done.  Also, I did a couple minor refactorings that weren't strictly necessary but which simplified and shortened the code, so there.  Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @exarkun|
|-|-|

Trac ate my comment, but basically self.started = True in run is not tested, but it's probably not new here. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

Thanks.  You're right that it's not tested.  It's not quite new here; it was being set in trunk, but it was necessary before.  Now it's not necessary anymore.  I left it out of some idea of consistency (making sure it was always `True` by the time a request was done), but I probably should have deleted it.

Can you merge this for me?  Either leaving that line, or deleting it (assuming all the tests still pass, which I expect they will).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set status to `closed`|
|-|-|

(In [27354]) Merge wsgi-conn-lost-4025

Author: exarkun
Reviewer: therve
Fixes [#4025](https://github.com/twisted/twisted/issues/4025)

Prevent twisted.web wsgi wrapper for calling request.finish twice, for example
when the connection is lost in the middle of the request.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

I left the line for now, it didn't hurt.
